File: concepts/06_text_binary_and_serialization/04_binary_buffers_views_and_endianness/r/test_02_alignment_and_lifetime.R
# polyglot-family: text_binary_and_serialization
# polyglot-concept: binary_buffers_views_and_endianness
# polyglot-related: languages/r/tooling_and_runtime/16_native_interface_and_unsafe_boundaries/
# polyglot-related+: test_123_dot_call_sexp_types_and_allocations.R
#
# 共同问题：typed view 的对齐和底层缓冲区生命周期如何约束。
# 对照观察：base R raw vector 没有可写 typed view/alignment API；`readBin` 从任意字节偏移解码副本。

local({
    bytes <- as.raw(c(0xff, 0x01, 0x00, 0x00, 0x00))
    connection <- rawConnection(bytes, open = "rb")
    on.exit(close(connection), add = TRUE)
    seek(connection, 1L)
    value <- readBin(connection, integer(), n = 1L, size = 4L, endian = "little")

    stopifnot(
        identical(value, 1L),
        identical(typeof(bytes), "raw"),
        identical(length(bytes), 5L)
    )
})

File: concepts/06_text_binary_and_serialization/04_binary_buffers_views_and_endianness/julia/test_02_alignment_and_lifetime.jl
# polyglot-family: text_binary_and_serialization
# polyglot-concept: binary_buffers_views_and_endianness
# polyglot-related: languages/julia/tooling_and_runtime/16_ffi_pointers_and_unsafe_boundaries/
# polyglot-related+: test_123_cconvert_unsafe_convert_and_gc_preserve.jl
#
# 共同问题：typed view 对大小和对齐有什么要求；裸 pointer 的有效期由谁保证。
# 对照观察：reinterpret 要求字节尺寸可整除；unsafe pointer 操作由调用方用拥有对象和 GC.@preserve 约束。

using Test

@testset "typed bytes 和 pointer 需要显式前置条件" begin
    @test_throws ArgumentError reinterpret(UInt32, UInt8[1, 2, 3])
    storage = Ref{UInt32}(0x01020304)
    pointer = Base.unsafe_convert(Ptr{UInt32}, storage)
    GC.@preserve storage begin
        @test unsafe_load(pointer) == 0x01020304
        unsafe_store!(pointer, 0x05060708)
        @test storage[] == 0x05060708
    end
end
